File: 1570-final-prices-with-a-special-discount-in-a-shop/final-prices-with-a-special-discount-in-a-shop.cpp
class Solution {
public:
    vector<int> finalPrices(vector<int>& prices) {
        stack<int> sta;
        for(int i = 0;i<prices.size();i++){
            while(sta.size()>0 && prices[sta.top()] >= prices[i]){
                prices[sta.top()] -= prices[i];
                sta.pop();
            }
            sta.push(i);
        }
        return prices;
    }
};

File: 1570-final-prices-with-a-special-discount-in-a-shop/final-prices-with-a-special-discount-in-a-shop.java
class Solution {
    public int[] finalPrices(int[] prices) {
        Stack<Integer> sta = new Stack<>();
        for(int i = 0;i<prices.length;i++){
            while(sta.size()>=1 && prices[sta.peek()] >= prices[i]){
                prices[sta.pop()] -= prices[i];
            }
            sta.push(i);
        }
        return prices;
    }
}